### DISCUSSION TITLE
What might a more iterative change look like?

### DISCUSSION THREAD

**Discussion by bdarcus** [bdarcus/csl-next-attic#15] [Q&A] [answered]
One might reasonably say this isn't worth the hassle.

So what's the alternative, other than freezing CSL?

If we ended up going this route, [here was my list of changes in the 1.1 branch](https://discourse.citationstyles.org/t/upcoming-csl-meetup-context/1767/5#draft-v11-changes-3), that we could use to help figure out what we'd also want to include in such a `CSL 2.0`.

**Comment by bdarcus** [comment]
I think it's probably:

1. allow collections of `cs:macro` elements in standalone files.
2. allow multiple `cs:citation` (and maybe `cs-bibliography`) elements, with a required `mode` (or similar) attribute, which we've previously discussed over at the CSL repo
3. maybe, the `transform` attribute added in a few places.

On 2, `cs:citation` and `cs:bibliography` are (I realized through this exercise) elements that configure list formatting, the former inline and the latter block, with a set of function and parameter configuration attributes defining the default transformation behavior in descendant rendering nodes (and sorting rules).

So in order to get different transformation and formatting behavior in different contexts, you probably need multiple such elements (though I don't believe it's necessary for bibliographies, unless you consider multi-bibs?).

Those changes wouldn't be too invasive, and would allow new things. We'd [already identified](https://github.com/citation-style-language/schema/pull/419) 2 as an option.

But they probably could only go so far (which may be fine).

PS - the `cs:choose` structure may be a bit verbose, but something like it is necessary, and I don't see a good alternative ATM in XML.

------------

Also, on this:

```xml
 <citation et-al-min="3" 
           et-al-use-first="1" 
           disambiguate-add-year-suffix="true" 
           disambiguate-add-names="true" 
           disambiguate-add-givenname="true" 
           collapse="year" 
           givenname-disambiguation-rule="primary-name-with-initials">
```

There's some confused scope here. None of them really refer to the citation; they refer to formatting:

1. the list of author/creator names
2. individual item data
3. disambiguation of author-year strings within a list of all such strings

The only way to make it make sense is to think of these as really default settings that apply to the child rendering, that currently I guess you can't locally override?

**Reply by bdarcus** [reply]
I wonder if an even more minimal change would be adding a new `cs:mode` conditional attribute, so you could do:

```xml
<when mode="intext">....
```

Maybe it's worth testing this.

**Reply by denismaier** [reply]
> The only way to make it make sense is to think of these as really default settings that apply to the child rendering, that currently I guess you can't locally override?

At least according to the specs, this is how that should be understood. Default settings that should be possible to override by an element lower in the tree.

**Reply by denismaier** [reply]
> I wonder if an even more minimal change would be adding a new `cs:mode` conditional attribute, so you could do:
> 
> ```
> <when mode="intext">....
> ```
> 
> Maybe it's worth testing this.

Where would that `when` clause appear?

**Reply by bdarcus** [reply]
See #16, which I posted a bit ago; option C.

EDIT:

```xml
    <if mode="in-text">
      <text macro="author-intext" suffix=" "/>
      <group prefix="(" suffix=")" delimiter="; ">
        <group delimiter=", ">
          <text macro="date-intext"/>
          <text macro="citation-locator"/>
        </group>
      </group>
    </if>
    <else>
      <group prefix="(" suffix=")" delimiter="; ">
        <group delimiter=", ">
          <text macro="author-paren"/>
          <text macro="date-intext"/>
          <text macro="citation-locator"/>
        </group>
      </group>
    </else>
```

**Reply by bdarcus** [reply]
A few things:

1. I do think this last option is the best of the "iterative" ones; the most flexible, and least intrusive, and could be done (as above) in a way that style compatibility would remain so long as processors know what to do with unknown conditional attributes. 
2. So I closed my original PR over at the main repo https://github.com/citation-style-language/schema/pull/419
3. The v1.1. branch is a mess and it's going to be a big hassle to fix it https://github.com/citation-style-language/schema/issues/429

I think the experiments and discussion so far have helped me better understand our options, but it strikes me it will still come down, in particular, talking with processor developers, probably not generally, but specifically.

Like, "if we add this attribute, how would you handle it? what other options do you see?"

And fixing that branch, such that ideally even it can auto-generate an accurate changelog, would likely facilitate that.

I think that, along with the pause in citeproc-rs, may suggest delaying wider discussion a bit. It could be even that a reasonable iterative approach would mean suggesting a couple of new attributes in existing CSL.

PS - I sent you a collaborator link, @denismaier, mostly so I get auto-complete on comments!

**Comment by bdarcus** [comment]
Just a few more high-level notes after thinking about this more:

1. The existing `cs:layout`, `cs:locators`, and `cs:names` (and probably `cs:group`) are examples of list formatting configuration; sometimes inline, and sometimes block.
2. A lot of attributes in those places are dedicated processing/transform arguments, that implementations can directly map to their internal functions.
3. Attributes are cheap to add, if we can say in the spec something like the below:

 > Processors shall ignore unknown attributes, and return the string as is. When in the context of conditional statements, the processor shall consider the result `nil`.

But the devil is in the details; not sure if we could cleanly integrate some strategy there based on this observation.

I think the practical question is if we can re-implement [split titles](https://github.com/citation-style-language/schema/pull/375) and narrative citations in a possibly more general way, using some of the ideas I've suggested here, but with minimal or zero new elements.

I just don't have the bandwidth to think about that more deeply than that.

Thoughts?

**Comment by bdarcus** [comment, marked as answer]
The high-level and tentative answer is probably some mix of:

1. allowing foreign attributes and attribute values in places
2. some kind of change to `cs:layout`, and possibly `cs:choose`, to allow greater flexibility, including ...
3. add `@cs:mode`

**Reply by denismaier** [reply]
That sound's reasonable.
